// @elijahjcobb/collections/dist/index.js
"use strict";
/**
 *
 * Elijah Cobb
 * elijah@elijahcobb.com
 * https://elijahcobb.com
 *
 *
 * Copyright 2019 Elijah Cobb
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
 * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
 * OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
var ECArray_1 = require("./array/ECArray");
exports.ECArray = ECArray_1.ECArray;
var ECArrayList_1 = require("./array/ECArrayList");
exports.ECArrayList = ECArrayList_1.ECArrayList;
var ECCSV_1 = require("./ECCSV");
exports.ECCSV = ECCSV_1.ECCSV;
var ECDictionary_1 = require("./map/ECDictionary");
exports.ECDictionary = ECDictionary_1.ECDictionary;
var ECEnum_1 = require("./ECEnum");
exports.ECEnum = ECEnum_1.ECEnum;
var ECIterator_1 = require("./ECIterator");
exports.ECIterator = ECIterator_1.ECIterator;
var ECMap_1 = require("./map/ECMap");
exports.ECMap = ECMap_1.ECMap;
var ECPrototype_1 = require("./ECPrototype");
exports.ECPrototype = ECPrototype_1.ECPrototype;
var ECQueue_1 = require("./ECQueue");
exports.ECQueue = ECQueue_1.ECQueue;
var ECSet_1 = require("./ECSet");
exports.ECSet = ECSet_1.ECSet;
var ECStack_1 = require("./ECStack");
exports.ECStack = ECStack_1.ECStack;
